README.md
# EJ Selector Button
A flutter widget that works like dropdown button, except that instead of opening the dropdown to select items, it opens a dialog. Also, You can customize its button and items.
Or you can show dialog for selecting item using `showEJDialog`.

## Usage
You can customize `EJSelectorButton` or use its string factory which is less customizable and easier to use.

<img src="https://i.imgur.com/GN9c96G.gif" width="300" height="540">

```dart
EJSelectorButton.string(
  items: List.generate(10, (index) => 'item $index'),
  hint: 'Choose',
  useValue: false,
  divider: Divider(),
  textStyle: TextStyle(fontSize: 18),
  suffix: Icon(Icons.arrow_drop_down),
)
```


<img src="https://i.imgur.com/vVx7uAF.gif" width="300" height="540">

```dart
final items = <ItemModel>[
  ItemModel(1, 'First Item'),
  ItemModel(2, 'Second Item'),
  ItemModel(3, 'Third Item'),
  ItemModel(4, 'Forth Item'),
  ItemModel(5, 'Fifth Item'),
];

EJSelectorButton<ItemModel>(
 useValue: false,
 hint: Text(
    'Click to choose',
    style: TextStyle(fontSize: 16, color: Colors.black),
 ),
 buttonBuilder: (child, value) => Container(
  alignment: Alignment.center,
  height: 60,
  width: 150,
  decoration: BoxDecoration(
  borderRadius: BorderRadius.circular(4),
  color: Colors.white,
 ),
 child: value != null
      ? Text(
         value.name,
         style: TextStyle(fontSize: 16, color: Colors.black),
        )
      : child,
 ),
 selectedWidgetBuilder: (valueOfSelected) => Container(
  padding: const EdgeInsets.symmetric(vertical: 16, horizontal: 32),
  child: Text(
     valueOfSelected.name,
     style: TextStyle(fontSize: 20, color: Colors.blue),
    ),
 ),
 items: items
     .map(
       (item) => EJSelectorItem(
        value: item,
        widget: Container(
        padding:
           const EdgeInsets.symmetric(vertical: 16, horizontal: 32),
        child: Text(
         item.name,
         style: TextStyle(fontSize: 16),
        ),
      ),
     ),
    ).toList(),
)

class ItemModel {
  ItemModel(this.id, this.name);

  final int id;
  final String name;
}
```

### Using as Function
```dart
final s = await showEJDialog<int>(
     context: context,
     selected: selectedId,
     selectedWidgetBuilder: (selectedId) => Container(
     padding: const EdgeInsets.symmetric(vertical: 16, horizontal: 32),
     child: Text(
     items.firstWhere((item) => item.id == selectedId).name,
     style: TextStyle(fontSize: 20, color: Colors.blue),
    ),
 ),
    items: items
        .map(
          (item) => EJSelectorItem(
             value: item.id,
             widget: Container(
             padding: const EdgeInsets.symmetric(
             vertical: 16, horizontal: 32),
             child: Text(
                item.name,
                style: TextStyle(fontSize: 16),
               ),
            ),
         ),
      ) .toList(),
);
if (s != null) {
     setState(() {
     selectedId = s.value;
    });
 }
```

CHANGELOG.md
# [1.0.3]

- Fixed issues from flutter static analysis

## [1.0.2+1]

- Fixed issues from flutter static analysis

## [1.0.2]

- Fixed issues from flutter static analysis

## [1.0.1]

- Added `disable` and `onDisableTap` parameters

## [1.0.0]

- Migrate to null-safety

## [0.1.1+1]

- Updated documentation comments

## [0.1.1]

- Added `buttonTextOverFlow` parameter to `string` factory

## [0.1.0]

- Added `showEJDialog` method
- Updated README.md

## [0.0.3+1]

- Updated description and readme

## [0.0.3]

- Added `divider` parameter
- Fixed documentation
- Added example

## [0.0.2]

- Fixed item background color
- Removed dialog shape

## [0.0.1]

- Initial Release


lib/src/selector_button.dart
import 'package:flutter/material.dart';

import './util/extensions.dart';

part 'selector_dialog.dart';

class EJSelectorButton<T> extends StatefulWidget {
  EJSelectorButton({
    Key? key,
    required this.items,
    this.value,
    this.useValue = true,
    this.hint,
    this.dialogWidth = 80,
    this.dialogHeight,
    this.buttonBuilder,
    this.selectedWidgetBuilder,
    this.onChange,
    this.divider,
    this.onTap,
    this.alwaysShowScrollBar,
    this.disable = false,
    this.onDisableTap,
  })  : assert(
          items.isEmpty ||
              value == null ||
              items.where((item) => item.value == value).length == 1,
          "There should be exactly one item with [item]'s value: "
          '$value.\n'
          'Either zero or 2 or more [item]s were detected '
          'with the same value',
        ),
        assert(T.runtimeType != Widget,
            '{EJSelectorButton}\'s type can\'t be {Widget}'),
        super(key: key);

  /// The list of items the user can select and it's can't be null.
  final List<EJSelectorItem<T>> items;

  /// The value of the currently selected [EJSelectorItem].
  ///
  /// [EJSelectorButton] will use [value] if [useValue] is true;
  /// If [value] is null and [useValue] is true and [hint] is non-null,
  /// the [hint] widget is displayed as a placeholder for
  /// the [EJSelectorButton]'s value.
  final T? value;

  /// Specifies whether to use the value.
  ///
  /// defaults to true.
  final bool useValue;

  /// A placeholder widget that is displayed by the button.
  ///
  /// If [value] is null, this widget is displayed as a placeholder for
  /// the [EJSelectorButton]'s value.
  final Widget? hint;

  /// The Height of the dialog which shows for select items.
  ///
  /// [dialogHeight] must be lower than screen height - 200.
  /// If [dialogHeight] is more than screen height - 200, height of the dialog
  /// will be equal to screen height - 200.
  /// If [dialogHeight] is null, height of the dialog will be fit to
  /// height of its children.
  final double? dialogHeight;

  /// The Width of the dialog which shows for select items.
  ///
  /// [dialogWidth] must be lower than screen width - 200.
  /// If [dialogWidth] is more than screen width - 200, width of the dialog
  /// will be equal to screen width - 200.
  ///
  /// defaults to 80.
  final double dialogWidth;

  /// Custom builder for the main widget of [EJSelectorButton].
  ///
  /// [buttonBuilder]'s child will be widget of selected item
  /// and it's value will be value of selected item, if no item is selected,
  /// child will be hint and if hint is null, child will be [Container].
  ///
  /// If [buttonBuilder] is null or return null,
  /// main widget of [EJSelectorButton] will be widget of selected item and
  /// if no item is selected, it will be hint.
  final Widget Function(Widget child, T? value)? buttonBuilder;

  /// Custom builder for selected item in dialog.
  final Widget Function(T valueOfSelected)? selectedWidgetBuilder;

  /// Called when the user selects an item.
  final void Function(T value)? onChange;

  /// Called when the button is tapped.
  ///
  /// This is distinct from [onChanged], which is called when the user
  /// selects an item from the dialog.
  final VoidCallback? onTap;

  /// Custom builder for divider between items in the dialog.
  final Widget? divider;

  /// Indicates whether the scrollbar for the dialog should always be visible.
  final bool? alwaysShowScrollBar;

  /// Disables the selector.
  ///
  /// defaults to false.
  final bool disable;

  /// Called when the button is tapped and [disable] sets on true.
  final VoidCallback? onDisableTap;

  static EJSelectorButton<String> string({
    Key? key,
    required List<String> items,
    String? value,
    bool useValue = true,
    String? hint,
    double dialogWidth = 80,
    double? buttonWidth,
    double? buttonHeight,
    double itemExtent = 50,
    void Function(String)? onChange,
    VoidCallback? onTap,
    Widget? divider,
    bool underline = true,
    TextStyle? textStyle,
    TextStyle? dialogTextStyle,
    TextStyle? hintStyle,
    Widget? suffix,
    Widget? prefix,
    TextOverflow? buttonTextOverFlow,
    Color underlineColor = Colors.grey,
    BoxDecoration? decoration,
    Alignment alignment = Alignment.center,
    EdgeInsets padding = const EdgeInsets.symmetric(horizontal: 8, vertical: 4),
    EdgeInsets margin = EdgeInsets.zero,
    bool alwaysShownScrollbar = false,
    bool disable = false,
    VoidCallback? onDisableTap,
  }) {
    return EJSelectorButton<String>(
      hint: Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          if (prefix != null) prefix,
          Container(
            alignment: alignment,
            child: Text(hint ?? '', style: hintStyle ?? textStyle),
          ).addExpanded(buttonWidth != null),
          if (suffix != null) suffix,
        ],
      ),
      items: List.generate(items.length, (index) {
        final item = items[index];
        return EJSelectorItem<String>(
            value: item,
            widget: Builder(
              builder: (context) => Container(
                height: itemExtent,
                alignment: Alignment.center,
                child: Text(
                  item,
                  style: dialogTextStyle ??
                      textStyle ??
                      Theme.of(context).textTheme.bodyMedium,
                ),
              ),
            ));
      }),
      disable: disable,
      onDisableTap: onDisableTap,
      alwaysShowScrollBar: alwaysShownScrollbar,
      useValue: useValue,
      value: value,
      selectedWidgetBuilder: (value) => Builder(
        builder: (context) => Container(
          height: itemExtent,
          alignment: Alignment.center,
          child: Text(
            value,
            style: Theme.of(context).textTheme.headlineSmall!.copyWith(
                color: Theme.of(context).colorScheme.secondary,
                fontWeight: FontWeight.bold),
          ),
        ),
      ),
      divider: divider,
      buttonBuilder: (child, value) => Builder(
        builder: (context) => Container(
          height: buttonHeight,
          width: buttonWidth,
          padding: padding,
          margin: margin,
          alignment: alignment,
          decoration: decoration ??
              BoxDecoration(
                border: underline
                    ? Border(
                        bottom: BorderSide(color: underlineColor, width: 2))
                    : Border(),
              ),
          child: value != null
              ? Row(
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    if (prefix != null) prefix,
                    Container(
                      alignment: alignment,
                      child: Text(
                        value,
                        overflow: buttonTextOverFlow,
                        style:
                            textStyle ?? Theme.of(context).textTheme.bodyMedium,
                      ),
                    ).addExpanded(buttonWidth != null),
                    if (suffix != null) suffix,
                  ],
                )
              : child,
        ),
      ),
      onChange: onChange,
      onTap: onTap,
      dialogWidth: dialogWidth,
      dialogHeight: items.length * itemExtent,
      key: key,
    );
  }

  @override
  _EJSelectorButtonState<T> createState() => _EJSelectorButtonState<T>();
}

class _EJSelectorButtonState<T> extends State<EJSelectorButton<T>> {
  EJSelectorItem<T>? _selected;

  @override
  void initState() {
    if (widget.value != null) {
      _selected = widget.items.firstWhere((item) => item.value == widget.value);
    }
    super.initState();
  }

  @override
  void didUpdateWidget(covariant EJSelectorButton<T> oldWidget) {
    if (widget.useValue) {
      if (widget.value != null) {
        _selected =
            widget.items.firstWhere((item) => item.value == widget.value);
      } else {
        _selected = null;
      }
    } else if (widget.value != null) {
      _selected = widget.items.firstWhere((item) => item.value == widget.value);
    }
    super.didUpdateWidget(oldWidget);
  }

  @override
  Widget build(BuildContext context) {
    final child = _selected?.widget ?? widget.hint ?? Container();
    final button = widget.buttonBuilder != null
        ? widget.buttonBuilder!(child, _selected?.value)
        : child;

    return button.mouseRegion(
        cursor: widget.items.isNotEmpty || widget.onTap != null
            ? SystemMouseCursors.click
            : SystemMouseCursors.basic,
        onTap: widget.disable
            ? widget.onDisableTap
            : () async {
                EJSelectorItem<T>? s;
                if (widget.items.isNotEmpty) {
                  s = await showEJDialog<T>(
                    context: context,
                    barrierDismissible: true,
                    selected: _selected?.value,
                    alwaysShownScrollbar: widget.alwaysShowScrollBar ?? true,
                    selectedWidgetBuilder: widget.selectedWidgetBuilder,
                    divider: widget.divider,
                    dialogHeight: widget.dialogHeight,
                    dialogWidth: widget.dialogWidth,
                    items: widget.items,
                  );
                }
                if (s != null) {
                  setState(() {
                    _selected = s;
                  });
                  if (widget.onChange != null) {
                    widget.onChange!(s.value);
                  }
                }
                if (widget.onTap != null) {
                  widget.onTap!();
                }
              });
  }
}

class EJSelectorItem<T> {
  EJSelectorItem({
    required this.value,
    required this.widget,
  });

  /// The value of item.
  final T value;

  /// The widget of item.
  final Widget widget;
}


lib/src/selector_dialog.dart
part of 'selector_button.dart';

class _Dialog<T> extends StatefulWidget {
  const _Dialog({
    Key? key,
    required this.selected,
    required this.items,
    required this.selectedWidgetBuilder,
    required this.alwaysShownScrollbar,
    this.divider,
    this.dialogHeight,
    this.dialogWidth,
  }) : super(key: key);

  final T? selected;
  final List<EJSelectorItem<T>> items;
  final Widget Function(Widget child, T value) selectedWidgetBuilder;
  final Widget? divider;
  final double? dialogHeight;
  final double? dialogWidth;
  final bool alwaysShownScrollbar;

  @override
  _DialogState<T> createState() => _DialogState<T>();
}

class _DialogState<T> extends State<_Dialog<T>> {
  ScrollController? _controller;
  late List<GlobalKey> _keys;
  late List<Widget> _children;
  int? _selectedItemIndex;

  @override
  void initState() {
    _controller = ScrollController();
    _keys = List<GlobalKey>.generate(widget.items.length, (i) => GlobalKey());
    final hasDivider = widget.divider != null;
    _children = List.generate(
        hasDivider ? widget.items.length * 2 - 1 : widget.items.length,
        (index) {
      if (hasDivider && index % 2 != 0) {
        return widget.divider!;
      }

      final itemIndex = hasDivider ? (index / 2).floor() : index;
      final EJSelectorItem<T> item = widget.items[itemIndex];
      final isSelected =
          widget.selected != null && item.value == widget.selected;
      if (isSelected) {
        _selectedItemIndex = itemIndex;
      }

      final child = isSelected
          ? widget.selectedWidgetBuilder(item.widget, item.value)
          : item.widget;

      return Material(
        key: _keys[itemIndex],
        color: Colors.transparent,
        child: InkWell(
          onTap: () {
            Navigator.pop(context, item);
          },
          child: child,
        ),
      );
    });
    if (_selectedItemIndex != null) {
      WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
        Scrollable.ensureVisible(
          _keys[_selectedItemIndex!].currentContext!,
          curve: Curves.fastOutSlowIn,
          duration: Duration(milliseconds: 200),
        );
      });
    }
    super.initState();
  }

  @override
  void dispose() {
    _controller?.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) => LayoutBuilder(
        builder: (_, constraints) => SimpleDialog(
          children: [
            Container(
              width: widget.dialogWidth != null &&
                      (widget.dialogWidth! < (constraints.maxWidth - 200))
                  ? widget.dialogWidth
                  : (constraints.maxWidth - 200),
              height: widget.dialogHeight != null
                  ? (widget.dialogHeight! < (constraints.maxHeight - 200)
                      ? widget.dialogHeight
                      : (constraints.maxHeight - 200))
                  : null,
              child: Scrollbar(
                controller: _controller,
                thumbVisibility: widget.alwaysShownScrollbar,
                child: SingleChildScrollView(
                  controller: _controller,
                  child: Column(
                    children: _children,
                  ),
                ),
              ),
            ),
          ],
        ),
      );
}

Future<EJSelectorItem<T>?> showEJDialog<T>({
  required BuildContext context,
  required List<EJSelectorItem<T>> items,
  T? selected,
  bool alwaysShownScrollbar = true,
  Widget Function(T valueOfSelected)? selectedWidgetBuilder,
  Widget? divider,
  double? dialogHeight,
  double dialogWidth = 80,
  bool barrierDismissible = true,
  Color barrierColor = Colors.black54,
  String? barrierLabel,
  bool useSafeArea = true,
  bool useRootNavigator = true,
  RouteSettings? routeSettings,
}) async {
  return showDialog<EJSelectorItem<T>>(
    context: context,
    barrierDismissible: barrierDismissible,
    useRootNavigator: useRootNavigator,
    routeSettings: routeSettings,
    useSafeArea: useSafeArea,
    barrierColor: barrierColor,
    barrierLabel: barrierLabel,
    builder: (context) => _Dialog<T>(
      selected: selected,
      alwaysShownScrollbar: alwaysShownScrollbar,
      selectedWidgetBuilder: (child, value) {
        if (selectedWidgetBuilder != null) {
          return selectedWidgetBuilder(value!);
        } else {
          return child;
        }
      },
      divider: divider,
      dialogHeight: dialogHeight,
      dialogWidth: dialogWidth,
      items: items,
    ),
  );
}


lib/src/util/extensions.dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

extension MouseExtension on Widget {
  Widget mouseRegion({
    void Function(PointerHoverEvent)? onHover,
    void Function(PointerEnterEvent)? onEnter,
    void Function(PointerExitEvent)? onExit,
    void Function()? onTap,
    void Function()? onDoubleTap,
    void Function()? onLongPress,
    void Function()? onTapCancel,
    void Function(TapDownDetails)? onTapDown,
    InkWell? inkWell,
    MouseCursor cursor = SystemMouseCursors.click,
  }) =>
      MouseRegion(
        cursor: cursor,
        onEnter: onEnter,
        onExit: onExit,
        onHover: onHover,
        child: onTap != null
            ? inkWell != null
                ? Material(
                    color: Colors.white,
                    child: InkWell(
                      onTap: onTap,
                      onDoubleTap: onDoubleTap,
                      onLongPress: onLongPress,
                      onTapCancel: onTapCancel,
                      onTapDown: onTapDown,
                      focusColor: inkWell.focusColor,
                      hoverColor: inkWell.hoverColor,
                      highlightColor: inkWell.highlightColor,
                      splashColor: inkWell.splashColor,
                      radius: inkWell.radius,
                      borderRadius: inkWell.borderRadius,
                      focusNode: inkWell.focusNode,
                      autofocus: inkWell.autofocus,
                      key: inkWell.key,
                      excludeFromSemantics: inkWell.excludeFromSemantics,
                      canRequestFocus: inkWell.canRequestFocus,
                      customBorder: inkWell.customBorder,
                      enableFeedback: inkWell.enableFeedback,
                      onFocusChange: inkWell.onFocusChange,
                      onHighlightChanged: inkWell.onHighlightChanged,
                      overlayColor: inkWell.overlayColor,
                      splashFactory: inkWell.splashFactory,
                      child: this,
                    ),
                  )
                : GestureDetector(
                    onTap: onTap,
                    onDoubleTap: onDoubleTap,
                    onLongPress: onLongPress,
                    onTapCancel: onTapCancel,
                    onTapDown: onTapDown,
                    child: this,
                  )
            : this,
      );

  Widget addExpanded(bool shouldExpand) =>
      shouldExpand ? Expanded(child: this) : this;
}


example/lib/main.dart
import 'package:ej_selector/ej_selector.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Selector Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        shadowColor: Colors.grey,
        dialogTheme: DialogTheme(
          elevation: 2,
          shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(8),
          ),
        ),
        brightness: Brightness.dark,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  var selectedIndex = 0;

  /// You can use [EJSelector] in tree ways:
  ///
  /// 1. use [EJSelectorButton.string].
  /// 2. use [EJSelectorButton] which you can customize.
  /// 3. use [showEJDialog] function.

  @override
  Widget build(BuildContext context) {
    final childList = <Widget>[
      StringSelector(),
      StringSelectorUsingValue(),
      CustomSelector(),
      CustomSelectorUsingValue(),
      SelectorUsingFunction(),
    ];
    return DefaultTabController(
      length: 5,
      initialIndex: 0,
      child: Scaffold(
        appBar: AppBar(
          title: Text('Selector Example'),
          bottom: TabBar(
            isScrollable: true,
            labelPadding:
                const EdgeInsets.symmetric(vertical: 8, horizontal: 16),
            tabs: [
              Text(
                'String',
                textAlign: TextAlign.center,
                style: TextStyle(fontSize: 16),
              ),
              Text(
                'String using value',
                textAlign: TextAlign.center,
                style: TextStyle(fontSize: 16),
              ),
              Text(
                'Custom',
                textAlign: TextAlign.center,
                style: TextStyle(fontSize: 16),
              ),
              Text(
                'Custom using value',
                textAlign: TextAlign.center,
                style: TextStyle(fontSize: 16),
              ),
              Text(
                'Using function',
                textAlign: TextAlign.center,
                style: TextStyle(fontSize: 16),
              ),
            ],
            onTap: (index) {
              selectedIndex = index;
              setState(() {});
            },
          ),
        ),
        body: Row(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                childList[selectedIndex],
              ],
            ),
          ],
        ),
      ),
    );
  }
}

class StringSelector extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return EJSelectorButton.string(
      items: List.generate(10, (index) => 'item $index'),
      hint: 'Choose',
      useValue: false,
      divider: Divider(),
      textStyle: TextStyle(fontSize: 18),
      suffix: Icon(Icons.arrow_drop_down),
    );
  }
}

class StringSelectorUsingValue extends StatefulWidget {
  @override
  _StringSelectorUsingValueState createState() =>
      _StringSelectorUsingValueState();
}

class _StringSelectorUsingValueState extends State<StringSelectorUsingValue> {
  String? value;

  @override
  Widget build(BuildContext context) {
    return EJSelectorButton.string(
      items: List.generate(10, (index) => 'item $index'),
      value: value,
      onChange: (newValue) {
        value = newValue;
        setState(() {});
      },
      hint: 'Choose',
      divider: Divider(),
      textStyle: TextStyle(fontSize: 18),
      suffix: Icon(Icons.arrow_drop_down),
    );
  }
}

class CustomSelector extends StatelessWidget {
  final items = <ItemModel>[
    ItemModel(1, 'First Item'),
    ItemModel(2, 'Second Item'),
    ItemModel(3, 'Third Item'),
    ItemModel(4, 'Forth Item'),
    ItemModel(5, 'Fifth Item'),
  ];

  @override
  Widget build(BuildContext context) {
    return EJSelectorButton<ItemModel>(
      useValue: false,
      hint: Text(
        'Click to choose',
        style: TextStyle(fontSize: 16, color: Colors.black),
      ),
      buttonBuilder: (child, value) => Container(
        alignment: Alignment.center,
        height: 60,
        width: 150,
        decoration: BoxDecoration(
          borderRadius: BorderRadius.circular(4),
          color: Colors.white,
        ),
        child: value != null
            ? Text(
                value.name,
                style: TextStyle(fontSize: 16, color: Colors.black),
              )
            : child,
      ),
      selectedWidgetBuilder: (valueOfSelected) => Container(
        padding: const EdgeInsets.symmetric(vertical: 16, horizontal: 32),
        child: Text(
          valueOfSelected.name,
          style: TextStyle(fontSize: 20, color: Colors.blue),
        ),
      ),
      items: items
          .map(
            (item) => EJSelectorItem(
              value: item,
              widget: Container(
                padding:
                    const EdgeInsets.symmetric(vertical: 16, horizontal: 32),
                child: Text(
                  item.name,
                  style: TextStyle(fontSize: 16),
                ),
              ),
            ),
          )
          .toList(),
    );
  }
}

class CustomSelectorUsingValue extends StatefulWidget {
  @override
  _CustomSelectorUsingValueState createState() =>
      _CustomSelectorUsingValueState();
}

class _CustomSelectorUsingValueState extends State<CustomSelectorUsingValue> {
  final items = <ItemModel>[
    ItemModel(1, 'First Item'),
    ItemModel(2, 'Second Item'),
    ItemModel(3, 'Third Item'),
    ItemModel(4, 'Forth Item'),
    ItemModel(5, 'Fifth Item'),
  ];
  int? selectedId;

  @override
  Widget build(BuildContext context) {
    return EJSelectorButton<int>(
      value: selectedId,
      onChange: (id) {
        selectedId = id;
        setState(() {});
      },
      hint: Text(
        'Click to choose',
        style: TextStyle(fontSize: 16, color: Colors.black),
      ),
      buttonBuilder: (child, id) => Container(
        alignment: Alignment.center,
        height: 60,
        width: 150,
        decoration: BoxDecoration(
          borderRadius: BorderRadius.circular(4),
          color: Colors.white,
        ),
        child: id != null
            ? Text(
                items.firstWhere((item) => item.id == id).name,
                style: TextStyle(fontSize: 16, color: Colors.black),
              )
            : child,
      ),
      selectedWidgetBuilder: (selectedId) => Container(
        padding: const EdgeInsets.symmetric(vertical: 16, horizontal: 32),
        child: Text(
          items.firstWhere((item) => item.id == selectedId).name,
          style: TextStyle(fontSize: 20, color: Colors.blue),
        ),
      ),
      items: items
          .map(
            (item) => EJSelectorItem(
              value: item.id,
              widget: Container(
                padding:
                    const EdgeInsets.symmetric(vertical: 16, horizontal: 32),
                child: Text(
                  item.name,
                  style: TextStyle(fontSize: 16),
                ),
              ),
            ),
          )
          .toList(),
    );
  }
}

class SelectorUsingFunction extends StatefulWidget {
  @override
  _SelectorUsingFunctionState createState() => _SelectorUsingFunctionState();
}

class _SelectorUsingFunctionState extends State<SelectorUsingFunction> {
  final items = <ItemModel>[
    ItemModel(1, 'First Item'),
    ItemModel(2, 'Second Item'),
    ItemModel(3, 'Third Item'),
    ItemModel(4, 'Forth Item'),
    ItemModel(5, 'Fifth Item'),
  ];
  int? selectedId;

  @override
  Widget build(BuildContext context) {
    return ElevatedButton(
      style: ElevatedButton.styleFrom(
          backgroundColor: Colors.white,
          padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 24)),
      onPressed: () async {
        final s = await showEJDialog<int>(
          context: context,
          selected: selectedId,
          selectedWidgetBuilder: (selectedId) => Container(
            padding: const EdgeInsets.symmetric(vertical: 16, horizontal: 32),
            child: Text(
              items.firstWhere((item) => item.id == selectedId).name,
              style: TextStyle(fontSize: 20, color: Colors.blue),
            ),
          ),
          items: items
              .map(
                (item) => EJSelectorItem(
                  value: item.id,
                  widget: Container(
                    padding: const EdgeInsets.symmetric(
                        vertical: 16, horizontal: 32),
                    child: Text(
                      item.name,
                      style: TextStyle(fontSize: 16),
                    ),
                  ),
                ),
              )
              .toList(),
        );
        if (s != null) {
          setState(() {
            selectedId = s.value;
          });
        }
      },
      child: Text(
        selectedId != null
            ? items.firstWhere((element) => element.id == selectedId).name
            : 'Click to choose',
        style: TextStyle(color: Colors.black),
      ),
    );
  }
}

class ItemModel {
  ItemModel(this.id, this.name);

  final int id;
  final String name;
}
